### PR TITLE
A7134_FMCZ project update [Draft]

### PR DESCRIPTION
**Pull request by adrimbarean**
This pull request is to get early feedback on the new code.
Only the last commit contains relevant changes at this point and is to be reviewed.

### CODE REVIEW

File: projects/ad713x_fmcz/ad713x_fmc.c [hunk 1]
```diff
@@ -79,126 +88,170 @@
...
+	spi_eng_msg_cmds[1] = TRANSFER_BYTES_R(1);
 
 	Xil_ICacheEnable();
 	Xil_DCacheEnable();
 
-	ad713x_init(&dev, ad713x_default_init_param);
+	ad713x_init(&ad713x_dev_1, &ad713x_init_param_1);
 
-	spi_eng_init(&desc, spi_eng_init_params);
+	mdelay(1000);
```
Comment (cristiantpop) [inline projects/ad713x_fmcz/ad713x_fmc.c:188]: If delay is needed after init probably it should be placed inside `ad713x_init` function

File: projects/ad713x_fmcz/ad713x_fmc.c [hunk 2]
```diff
@@ -79,126 +88,170 @@
...
+	ad713x_init(&ad713x_dev_2, &ad713x_init_param_2);
+
+	mdelay(1000);
+
+	spi_eng_offload_init(&spi_eng_desc, &spi_eng_inits);
+
+	/** For this example, only offload is supported */
+	msg = calloc(1, sizeof *msg);
 	if (!msg)
 		return -1;
```
Comment (cristiantpop) [inline projects/ad713x_fmcz/ad713x_fmc.c:201]: Return FAILURE

File: projects/ad713x_fmcz/ad713x_fmc.c [hunk 3]
```diff
@@ -79,126 +88,170 @@
...
 
+	spi_eng_set_transfer_length(spi_eng_desc, 32);
+
 	msg->spi_msg_cmds = malloc(sizeof(spi_eng_msg_cmds));
 	msg->spi_msg_cmds = spi_eng_msg_cmds;
 	msg->msg_cmd_len = ARRAY_SIZE(spi_eng_msg_cmds);
 	msg->rx_buf_addr = 0x800000;
 	msg->tx_buf_addr = 0xA000000;
+	msg->rx_buf = calloc((AD7134_FMC_CH_NO * AD7134_FMC_SAMPLE_NO),
```
Comment (cristiantpop) [inline projects/ad713x_fmcz/ad713x_fmc.c:210]: Check if calloc succeeded

File: projects/ad713x_fmcz/ad713x_fmc.c [hunk 4]
```diff
@@ -79,126 +88,170 @@
...
+			     sizeof(uint32_t));
 
-	/* Init the rx buffer with 0s */
-	for (i = 0; i < ARRAY_SIZE(msg->rx_buf); i++)
-		msg->rx_buf[i] = 0;
+	spi_eng_offload_load_msg(spi_eng_desc, msg);
+	spi_eng_transfer_multiple_msgs(spi_eng_desc,
```
Comment (cristiantpop) [inline projects/ad713x_fmcz/ad713x_fmc.c:214]: Check return values

File: projects/ad713x_fmcz/ad713x_fmc.c [hunk 5]
```diff
@@ -79,126 +88,170 @@
...
+	offload_data = (uint32_t*)((struct spi_desc_extra *)spi_eng_desc->extra)->
+		       rx_dma_startaddr;
+	for(i = 0; i < AD7134_FMC_SAMPLE_NO; i++) {
+		j = 0;
+		while(j < 8) {
+			*(offload_data + j) <<= 1;
+			*(offload_data + j) &= 0xffffff00;
+			*(offload_data + j) >>= 8;
+
+			data = lsb * ((int32_t)*(offload_data + j) & 0xFFFFFF);
```
Comment (cristiantpop) [inline projects/ad713x_fmcz/ad713x_fmc.c:236]: Replace:
`data = lsb * ((int32_t)*(offload_data + j) & 0xFFFFFF);`
for better understanding, with:
`data = (*(offload_data + j) & 0xFFFFFF) >> 10;`

File: projects/ad713x_fmcz/ad713x_fmc.c (hunk 1 above)
Comment (adrimbarean) [reply projects/ad713x_fmcz/ad713x_fmc.c:188]: Unfortunately, I am not sure the delay is needed. Per datasheet is not, but I have not tested without it so I would have to do that when I recover the board.

File: projects/ad713x_fmcz/ad713x_fmc.c (hunk 2 above)
Comment (adrimbarean) [reply projects/ad713x_fmcz/ad713x_fmc.c:201]: Will do.

File: projects/ad713x_fmcz/ad713x_fmc.c (hunk 3 above)
Comment (adrimbarean) [reply projects/ad713x_fmcz/ad713x_fmc.c:210]: Will do.

File: projects/ad713x_fmcz/ad713x_fmc.c (hunk 4 above)
Comment (adrimbarean) [reply projects/ad713x_fmcz/ad713x_fmc.c:214]: Will do.

File: projects/ad713x_fmcz/ad713x_fmc.c (hunk 5 above)
Comment (adrimbarean) [reply projects/ad713x_fmcz/ad713x_fmc.c:236]: Is this the same equation? It seems that it would introduce a full scale error of 4-4.096=96mV

**Comment by adrimbarean** [comment]
V2:
 - check function and dynamic allocation success;
 - update to new device driver;

**Pull request merged**